python/lsst/resources/gs.py

```diff
@@ -0,0 +1,279 @@
+        # try to remove that resource
+        if transfer == "move":
+            # Transactions do not work here
+            src.remove()
+
+    @contextlib.contextmanager
+    def open(
+        self,
+        mode: str = "r",
+        *,
+        encoding: Optional[str] = None,
+        prefer_file_temporary: bool = False,
+    ) -> Iterator[IO]:
+        # Docstring inherited
+        if "x" in mode:
+            if self.exists():
+                raise FileExistsError(f"File at {self} already exists.")
+            mode = mode.replace("x", "w")
+        # The GCS API does not support append or read/write modes so for
+        # those we use the base class implementation.
+        if "+" in mode or "a" in mode:
+            with super().open(mode, encoding=encoding, prefer_file_temporary=prefer_file_temporary) as buffer:
+                yield buffer
+        else:
+            with self.blob.open(mode, encoding=encoding) as buffer:
+                yield buffer
+            if "w" in mode and self.exists():
+                # Force the blob to be resynced.
+                self._blob = None
+
+    def walk(
+        self, file_filter: Optional[Union[str, re.Pattern]] = None
+    ) -> Iterator[Union[List, Tuple[ResourcePath, List[str], List[str]]]]:
+        # We pretend that GCS uses directories and files and not simply keys.
+        if not (self.isdir() or self.is_root):
+            raise ValueError(f"Can not walk a non-directory URI: {self}")
+
+        if isinstance(file_filter, str):
+            file_filter = re.compile(file_filter)
+
+        # Limit each query to a single "directory" to match os.walk
+        # We could download all keys at once with no delimiter and work
+        # it out locally but this could potentially lead to large memory
+        # usage for millions of keys. It will also make the initial call
+        # to this method potentially very slow. If making this method look
+        # like os.walk was not required, we could query all keys with
+        # pagination and return them in groups of 1000, but that would
+        # be a different interface since we can't guarantee we would get
+        # them all grouped properly across the 1000 limit boundary.
+        prefix = self.relativeToPathRoot if not self.is_root else ""
+        prefix_len = len(prefix)
+        dirnames: Set[str] = set()
+        filenames = []
+        files_there = False
+
+        blobs = self.client.list_blobs(self.bucket, prefix=prefix, delimiter="/")
+        for page in blobs.pages:
+            # "Sub-directories" turn up as prefixes in each page.
+            dirnames.update(dir[prefix_len:] for dir in page.prefixes)
+
+            # Files are reported for this "directory" only.
+            # The prefix itself can be included as a file because we write
+            # a zero-length file for mkdir(). These must be filtered out.
+            if file_filter is not None:
```

Comment:
Seems wordier than the S3 way of doing it (get all files, then filter if necessary).

Reply:
Yes, When I was copying from the S3 code I thought that it would be better to filter things out before creating the list if the list was going to be big.

Reply:
I've changed it back to the S3 way of making a big list first and then filtering it.